### ISSUE TITLE
Bug: keras.src.ops.normalize() produces -inf gradients for small non-zero float16 inputs for the torch backend

### ISSUE THREAD

**Issue by Johnhi19** [keras-team/keras#23546]
### Description
`keras.src.ops.normalize()`'s L2 path performs its squared-norm computation, epsilon clamp, and `rsqrt` entirely in the input's original dtype. For `float16` inputs, this causes the gradient to become non-finite (`inf`/`NaN`) even for inputs that are not close to zero, because `epsilon**2` underflows to `0.0` in `float16` and `rsqrt`'s derivative overflows `float16`'s range at magnitudes well within normal usage. The example below shows a gradient of `-inf` for an input array filled with `0.01`, versus a small finite gradient in `float32`.

### Reproduction 
```
import os
os.environ["KERAS_BACKEND"] = "torch"

import numpy as np
import torch

from keras.src.ops import normalize

x_np = np.full((4, 2), 0.01, dtype=np.float16) 

x = torch.tensor(x_np, requires_grad=True)
y = normalize(x, axis=-1, order=2)
y.sum().backward()
grad = x.grad.detach().cpu().numpy()

print(f"gradient:\n{grad}")
# gradient:
# [[-inf -inf]
#  [-inf -inf]
#  [-inf -inf]
#  [-inf -inf]]

```

#### Output with float32
```
gradient:
[[7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]]
```

#### Expected Behavior
I would expect the gradient not to be -inf because it is far from that. It would be more logical for the gradient to be 0 everywhere. 

### Possible Fix
A possible fix would be to upcast the input to at least float32 before computing the squared norm, epsilon clamp, and rsqrt, and only casting back to the original dtype at the end. The following shows an example patch

```
diff --git a/keras/src/ops/nn.py b/keras/src/ops/nn.py
index c4d1b396a..ad6d7fab0 100644
--- a/keras/src/ops/nn.py
+++ b/keras/src/ops/nn.py
@@ -2722,14 +2722,23 @@ def _normalize(x, axis=-1, order=2, epsilon=None):
     if 2 == order:
         # A special case: L2 normalization with `x * rsqrt(...)`
         # instead of `x / sqrt(...)`. Clamp the squared norm before the
         # rsqrt so zero vectors get a finite gradient.
+        original_dtype = getattr(backend, 'standardize_dtype', lambda d: d)(x.dtype)
+        compute_dtype = getattr(backend, 'result_type', lambda a, b: 'float32')(x.dtype, 'float32')
+        x_cast = backend.cast(x, dtype=compute_dtype)
         square_sum = backend.numpy.sum(
-            backend.numpy.square(x), axis=axis, keepdims=True
+            backend.numpy.square(x_cast), axis=axis, keepdims=True
         )
+        # compute epsilon squared in compute dtype to avoid underflow
+        eps_sq = backend.cast(epsilon * epsilon, dtype=compute_dtype)
         inv_norm = backend.math.rsqrt(
-            backend.numpy.maximum(square_sum, epsilon * epsilon)
+            backend.numpy.maximum(square_sum, eps_sq)
         )
-        return x * inv_norm
+        out = x_cast * inv_norm
+        return backend.cast(out, dtype=original_dtype)
+
     norm = backend.linalg.norm(x, ord=order, axis=axis, keepdims=True)
     denom = backend.numpy.maximum(norm, epsilon)
     return backend.numpy.divide(x, denom)
```

### AI Disclosure
I used an LLM based approach for research to find this bug and I also used an agentic approach to find a plausible fix. I reviewed the found bug by checking if the behavior is still active on the most current version and agreeing that this behavior is indeed unintended. I also reviewed the proposed patch and take full responsibility for it.

### Environment
- Keras version: 78d44013f at the master branch
- Backend: NumPy (`KERAS_BACKEND=numpy`)
- Python version: Python 3.11.2
- OS: Debian GNU/Linux 12 (bookworm)

**Comment by ChiragSW** [comment]
May I work on this?

**Comment by indrajeetapache** [comment]
## Analysis

Confirmed on current `master`: this is a real bug in the shared L2 path of `ops.normalize`, not torch-specific logic.

`_normalize` still does `square` / `epsilon**2` / `rsqrt` in the input dtype. For `float16`, that makes the `rsqrt` derivative overflow on ordinary magnitudes (e.g. `0.01`), so grads become `-inf`. The same values stay finite in `float32`.

Related history:
- #23075 / #23093 fixed **zero-vector NaN** by clamping before `rsqrt`, but did not land the half-precision upcast.
- #23449 is a different `rsqrt` subnormal issue and does not cover this.

Keras already uses the right pattern in `_rms_normalization` / layer norm: compute in at least `float32`, cast back. Expected behavior is finite grads matching the float32 path (not zeros).

Opening a PR that applies that upcast to the L2 normalize path and adds a float16 gradient test.
